lib/src/rules/prefer_collection_literals.dart

```diff
@@ -142,35 +142,35 @@ class _Visitor extends SimpleAstVisitor<void> {
 
   bool _shouldSkipLinkedHashLint(
       InstanceCreationExpression node, bool Function(DartType node) typeCheck) {
-    if (_isHashMap(node) || _isHashSet(node)) {
-      // Skip: LinkedHashSet<int> s =  ...; or LinkedHashMap<int> s =  ...;
-      var parent = node.parent;
-      if (parent is VariableDeclaration) {
-        var parent2 = parent.parent;
-        if (parent2 is VariableDeclarationList) {
-          var assignmentType = parent2.type?.type;
-          if (assignmentType != null && !typeCheck(assignmentType)) {
-            return true;
-          }
-        }
-      }
-      // Skip: function(LinkedHashSet()); when function(LinkedHashSet mySet) or
-      // function(LinkedHashMap()); when function(LinkedHashMap myMap)
-      if (parent is ArgumentList) {
-        final paramType = parent.arguments.first.staticParameterElement?.type;
-        if (paramType != null && !typeCheck(paramType)) {
-          return true;
-        }
-      }
-      // Skip: <int, LinkedHashSet>{}.putIfAbsent(3, () => LinkedHashSet());
-      // or <int, LinkedHashMap>{}.putIfAbsent(3, () => LinkedHashMap());
-      if (parent is ExpressionFunctionBody) {
-        var expressionType = parent.expression.staticType;
-        if (expressionType != null && !typeCheck(expressionType)) {
-          return true;
-        }
-      }
+    if (!_isHashMap(node) && !_isHashSet(node)) return false;
+    var parent = node.parent;
+    var staticType = _enforcedType(parent);
+    return staticType != null && !typeCheck(staticType);
+  }
+}
+
+/// Returns the static type which is pushed into an expression by it's parent.
+DartType /*?*/ _enforcedType(AstNode parent) {
```

Comment:
I might be completely misunderstanding what this method is trying to accomplish, but it looks like it's trying to find the type that's required by the context in which the `InstanceCreationExpression` appears. If so, then I think what we really need here is the algorithm implemented by the `_ContextTypeVisitor` in the `analysis_server` package, which is both more complete and more correct.

If I'm not understanding the purpose of this method then my comments below are probably also wrong.

Reply:
yes, that's exactly the goal. Is there a reasonable way we could expose that functionality here?

Specifically, the problem this was attempting to solve is that in some places if we _don't_ specifically use `LinkedHashMap` over `{}` then we'll get some other analysis error. We want to suppress the lint in only those places where honoring the lint would cause a static error.

Reply:
I suspect that we should just add a TODO and tackle that later. It will require moving the `_ContextTypeVisitor` to the `analyzer` package and exposing an API through which it can be used. That's a bigger task than I'm asking you to tackle here.